File: Peer/Application/Minidb.cs
﻿using Peer.Domain;
using System.Collections.Concurrent;
using System.Text.Json;

namespace Peer.Application;

public class Minidb
{
    public Minidb(string textPath, string filePath)
    {
        _textPath = textPath;
        _filePath = filePath;
        Load();
    }

    private readonly string _textPath, _filePath;
    private bool _isLockByCommit = false;
    private bool _isLockByShrink = false;
    private List<Data> _blockData = new();
    private ConcurrentDictionary<long, Block> _blocks = new();
    public long BlockIndex { get; private set; } = 0;
    public long BlockInBytes { get; private set; } = 0;
    public long SizeAllBlockInBytes { get; private set; } = 0;

    public long Write(long id, Message message)
    {
        string contentType = message.File?.ContentType ?? "";
        string fileName = message.File?.FileName ?? "";
        long fileSize = message.File?.Length ?? 0;
        long textSize = message.Text?.Length ?? 0;
        long querySize = textSize + fileSize;
        if (textSize > Config.MaxSizeText || querySize > Config.MaxSizeOneQuery || _blocks.ContainsKey(id))
        {
            return 0;
        }

        if (message.File != null)
        {
            using var stream = new FileStream(Path.Combine(_filePath, id.ToString() + Path.GetExtension(fileName)), FileMode.Create);
            message.File?.CopyTo(stream);
        }

        long addSecond = querySize > Config.LimitOtherSmallSizeOneQuery ? Config.LimitOtherBigMessageSecond : Config.LimitOtherSmallMessageSecond;
        if (SizeAllBlockInBytes <= Config.Limit1)
        {
            addSecond = querySize > Config.Limit1SmallSizeOneQuery ? Config.Limit1BigMessageSecond : Config.Limit1SmallMessageSecond;
        }
        long deleteUnixAt = new DateTimeOffset(DateTime.UtcNow).ToUnixTimeSeconds() + addSecond;
        _blockData.Add(new Data(id, message.Text, 0, fileName, deleteUnixAt, contentType, fileSize));
        _blocks[id] = new Block(BlockIndex, deleteUnixAt);
        SizeAllBlockInBytes += querySize;
        BlockInBytes += querySize;
        return deleteUnixAt;
    }

    public Data Get(long id)
    {
        for (int i = _blockData.Count - 1; i >= 0; i--)
        {
            if (_blockData[i].Id == id)
            {
                return _blockData[i];
            }
        }

        if (_blocks.TryGetValue(id, out Block? block))
        {
            string path = Path.Combine(_textPath, block.Index.ToString());
            if (!File.Exists(path)) return null;
            List<Data> datas = JsonSerializer.Deserialize<List<Data>>(ReadAllText(path)) ?? new List<Data>();
            return datas.FirstOrDefault(x => x.Id == id);
        }
        return null;
    }

    public void Shrink()
    {
        if (_isLockByShrink) return;
        _isLockByShrink = true;
        try
        {
            long nowUnix = new DateTimeOffset(DateTime.UtcNow).ToUnixTimeSeconds();
            HashSet<long> indexes = new();
            foreach (var block in _blocks.Values)
            {
                if (block.DeleteUnixAt < nowUnix)
                {
                    indexes.Add(block.Index);
                }
            }
            foreach (long index in indexes)
            {
                try
                {
                    string path = Path.Combine(_textPath, index.ToString());
                    if (!File.Exists(path)) continue;
                    List<Data> datas = JsonSerializer.Deserialize<List<Data>>(ReadAllText(path)) ?? new List<Data>();
                    List<long> deleteIds = new();
                    long sizeReduced = 0;
                    for (int i = datas.Count - 1; i >= 0; i--)
                    {
                        Data data = datas[i];
                        if (data.DeleteUnixAt < nowUnix)
                        {
                            string fullpath = Path.Combine(_filePath, data.Id.ToString() + Path.GetExtension(data.Filename ?? ""));
                            if (File.Exists(fullpath))
                            {
                                File.Delete(fullpath);
                            }
                            sizeReduced += data.Text.Length + data.FileSizeOfBytes;
                            deleteIds.Add(data.Id);
                            datas.RemoveAt(i);
                        }
                    }
                    if (datas.Count > 0)
                    {
                        WriteAllText(path, JsonSerializer.Serialize(datas));
                    }
                    else
                    {
                        File.Delete(path);
                    }
                    foreach (long id in deleteIds)
                    {
                        _blocks.TryRemove(id, out Block _);
                    }
                    SizeAllBlockInBytes -= sizeReduced;
                } catch { }
            }
        } catch { }
        _isLockByShrink = false;
    }

    public void Commit()
    {
        if (_isLockByCommit) return;
        _isLockByCommit = true;
        try
        {
            string path = Path.Combine(_textPath, BlockIndex.ToString());
            BlockIndex += 1;
            if (!File.Exists(path))
            {
                int blockCount = _blockData.Count;
                WriteAllText(path, JsonSerializer.Serialize(_blockData));
                _blockData.RemoveRange(0, blockCount);
                BlockInBytes = 0;
            }
        } catch { }
        _isLockByCommit = false;
    }

    public void Load()
    {
        foreach (string filePath in Directory.GetFiles(_textPath))
        {
            if (long.TryParse(filePath.Split('\\', '/').Last(), out long blockIndex))
            {
                List<Data> datas = JsonSerializer.Deserialize<List<Data>>(ReadAllText(filePath)) ?? new List<Data>();
                foreach (Data data in datas)
                {
                    if (data == null) continue;
                    _blocks[data.Id] = new Block(blockIndex, data.DeleteUnixAt);
                    SizeAllBlockInBytes += data.Text.Length + data.FileSizeOfBytes;
                }
                BlockIndex = blockIndex > BlockIndex ? blockIndex : BlockIndex;
            }
        }
        BlockIndex += 1;
    }

    private static void WriteAllText(string path, string? contents)
    {
        using (var stream = new FileStream(path, FileMode.Create, FileAccess.Write, FileShare.ReadWrite))
        using (var writer = new StreamWriter(stream))
        {
            writer.Write(contents);
        }
    }

    private static string ReadAllText(string path)
    {
        using (var stream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
        using (var reader = new StreamReader(stream))
        {
            return reader.ReadToEnd();
        }
    }
}

File: peerpy/application/minidb.py
import threading
import os
import time
import json
from config import Config
from domain.data import Data
from domain.block import Block
from domain.message import Message

class Minidb:
    def __init__(self, text_path: str, file_path: str, block_start_index: int = 0):
        self._text_path = text_path
        self._file_path = file_path
        self._is_lock_by_commit = False
        self._is_lock_by_shrink = False
        self._block_data = []
        self._blocks = {}
        self.block_index = block_start_index
        self.block_in_bytes = 0
        self.size_all_block_in_bytes = 0
        self.load()

    def write(self, message: Message) -> int:
        text_size = len(message.text)
        file_size = 0
        filename = ""
        content_type = ""
        if message.file is not None:
            file_size = message.file.content_length
            filename = message.file.filename;
            content_type = message.file.content_type
        query_size = text_size + file_size

        if text_size > Config.max_size_text or query_size > Config.max_size_one_query or message.id in self._blocks:
            return 0

        if message.file is not None:
            _, extension = os.path.splitext(filename)
            with open(os.path.join(self._file_path, str(message.id) + extension), 'wb') as stream:
                stream.write(message.file.read())

        add_second = Config.limit_other_big_message_second if query_size > Config.limit_other_small_size_one_query else Config.limit_other_small_message_second
        if (self.size_all_block_in_bytes <= Config.limit_1):
            add_second = Config.limit_1_big_message_second if query_size > Config.limit_1_small_size_one_query else Config.limit_1_small_message_second

        delete_unix_at = int(time.time()) + add_second
        self._block_data.append(Data(message.id, message.text, 0, filename, delete_unix_at, content_type, file_size))
        self._blocks[message.id] = Block(self.block_index, delete_unix_at)
        self.size_all_block_in_bytes += query_size
        self.block_in_bytes += query_size
        return delete_unix_at

    def get(self, id: int) -> Data:
        for data in self._block_data:
            if data.id == id:
                return data

        block = self._blocks.get(id)
        if block is not None:
            path = os.path.join(self._text_path, str(block.index))
            if os.path.exists(path):
                datas = []
                with open(path, 'r', encoding='utf-8') as f:
                    datas = [Data(**item) for item in json.load(f)]
                return next((x for x in datas if x.id == id), None)
        return None

    def load(self):
        for file_path in os.listdir(self._text_path):
            block_index = int(file_path)
            full_path = os.path.join(self._text_path, file_path)
            if os.path.isfile(full_path):
                datas = []
                try:
                    with open(full_path, 'r', encoding='utf-8') as f:
                        datas = [Data(**item) for item in json.load(f)]
                    for data in datas:
                        if not data == None:
                            self._blocks[data.Id] = Block(self.block_index, data.delete_unix_at);
                            self.size_all_block_in_bytes += len(data.text) + data.file_size_of_bytes
                    self.block_index = block_index if block_index > self.block_index else self.block_index
                except:
                    pass
        self.block_index += 1

    def commit(self):
        if not self._is_lock_by_commit:
            self._is_lock_by_commit = True
            try:
                path = os.path.join(self._text_path, str(self.block_index))
                self.block_index += 1;
                if not os.path.exists(path):
                    block_count = len(self._blocks)
                    json_string = json.dumps([data.to_dict() for data in self._block_data], ensure_ascii=True)
                    with open(path, 'w', encoding='utf-8') as f:
                        f.write(json_string);
                    self._block_data.clear()
                    self.block_in_bytes = 0
            except:
                pass
            self._is_lock_by_commit = False

    def shrink(self):
        if not self._is_lock_by_shrink:
            self._is_lock_by_shrink = True
            try:
                now_unix = int(time.time())
                indexes = set()
                for block in self._blocks.values():
                    if block.delete_unix_at < now_unix:
                        indexes.add(block.index)

                for index in indexes:
                    path = os.path.join(self._textPath, str(index))
                    if not os.path.exists(path): continue
                    datas = []
                    with open(path, 'r', encoding='utf-8') as f:
                        datas = [Data(**item) for item in json.load(f)]
                    delete_ids = []
                    size_reduced = 0
                    for i in range(datas.count):
                        data = datas[datas.count - 1 - i]
                        if data.delete_unix_at < now_unix:
                            _, extension = os.path.splitext(data.filename)
                            fullpath = os.path.join(self._file_path, str(data.id) + extension)
                            if os.path.exists(fullpath):
                                os.remove(fullpath)
                            size_reduced += len(data.text) + data.file_size_of_bytes
                            delete_ids.append(data.id)
                            datas.remove(data)

                    if datas.count > 0:
                        with open(path, 'w', encoding='utf-8') as f:
                            json.dump(datas, f, ensure_ascii=False, indent=0)
                    else:
                        os.remove(path)
                    for id in delete_ids:
                        self._blocks.pop(id, None)
                    self.size_all_block_in_bytes -= size_reduced
            except:
                pass
            self._is_lock_by_shrink = False
